### ISSUE TITLE
Implement a HadoopDataSource for Stratosphere that supports Hadoop Input formats

### ISSUE THREAD

**Issue by rmetzger** [stratosphere/stratosphere#382]
The goal of this task is to provide a generic data source that supports Hadoop Input formats.
The `HadoopDataSource` should convert the Hadoop Key/Values into Stratosphere's `Record` type.

Stratosphere API Example:

``` java
HadoopDataSource hd = new HadoopDataSource(ORCInputFormat.class, "hdfs:///file");       
// FileDataSource source = new FileDataSource(new TextInputFormat(), dataInput, "Input Lines");
MapOperator mapper = MapOperator.builder(new TokenizeLine())
    .input(hd)
    .name("Tokenize Lines")
    .build();
```

The `HadoopDataSource` must be generic, but I suggest to test it at least with (hadoop:) `AvroInputFormat`, `RCInputFormat` and `SequenceFileInputFormat`.
The `HadoopDataSource` must also support hadoop 1.x and 2.x at the same time (I hope that's possible). 

Spark has a similar feature http://spark.incubator.apache.org/docs/latest/api/core/index.html#org.apache.spark.SparkContext 

``` scala
def hadoopFile[K, V](path: String, inputFormatClass: Class[_ <: org.apache.hadoop.mapred.InputFormat[K,V]], keyClass: Class[K], valueClass: Class[V], minSplits: Int): RDD[(K, V)]
```

"Get an RDD for a Hadoop file with an arbitrary InputFormat"

One of the new students in the Stratosphere team ( @faisalmoeen ) is going to implement this.

**Comment by faisalmoeen** [comment]
Hadoop input formats are working fine with single node stratosphere now. My question is: if the splits are more than the max degree of parallelism, how does stratosphere behave. If I create more splits than the degree of parallelism=1 in local mode, only one split gets processed. Any ideas?

**Comment by rmetzger** [comment]
Cool! Thats good news.
I think the splits are just being processed one after another. So its not a problem if #splits > #dop. Lets see if @fhueske can say more about it.

**Comment by fhueske** [comment]
Splits are lazily assigned to tasks by the JobManager.
So tasks are created first and then splits are assigned to tasks that have no split. Once a task finishes processing a split, a new split is assigned. The assignment prefers to assign local splits as well.

**Comment by fhueske** [comment]
There seems to be a bug, if splits are not lazily assigned, i.e., only the first split is processed.
Does the job finish correctly (without error message, but not all data being processed)? Are actually multiple splits generated and managed by the JobManager? Is the InputFormat life cycle correctly implemented (see JavaDoc of InputFormat interface)?

I suggest to start with some debugging using the LocalExecutor. If that does not help, we can have a more detailed discussion.

**Comment by rmetzger** [comment]
I can also recommend to set the logging level to `DEBUG`

**Comment by faisalmoeen** [comment]
@fhueske 
Yes the job finishes correctly but without error message and not all data being processed.
Yes multiple splits are being generated
On checking the life cycle, I realized that the same instance of InputFormat is being used for multiple splits. I thought that the number of InputFormat instances was dependent on the number of InputSplits but this is not the case. It depends on the degree of Parallelism.  Lazy execution (as you mentioned), I did not consider this scenario in my code. I modified it and now it is working fine.
Thanks
